### ISSUE TITLE
Rewrite docs to teach to use the shell, e.g. not 'sbt compile'

### ISSUE THREAD

**Issue by dwijnand** [sbt/website#319]
It's far too common to hear "sbt is too slow" and then find out that the complainer doesn't use sbt's shell.

Eventually sbt server might solve this problem, but in the meantime we should start teaching to use the shell.

As an example, instead of saying "then run 'sbt compile'" we should say "run 'compile' in the sbt shell" or "start sbt and run 'compile'".

Credit to @adriaanm - it's originally his suggestion.

**Comment by eed3si9n** [comment]
During #306 (the January revamp), I have added the following in Getting Started (http://www.scala-sbt.org/0.13/docs/Running.html):

<img width="733" alt="running" src="https://cloud.githubusercontent.com/assets/184683/22624488/ef0ad86e-eb4a-11e6-9522-ee46135fa77e.png">

Coincidently @kmizu had also posted series of tweets in Japanese about this topic especially in comparison to Kotlin compilation speed, and wrote up a blog post http://qiita.com/kmizu/items/234a9e5ba91ee182a925. In his test, there's 2x difference in compiling from batch mode vs compiling from sbt shell.

We should continue to spread this knowledge and warn ppl about it.
There might also be a limitation to evangelism because many other build tools are executed from terminal, and it's hard to fight people's habit. In that sense, the approach sbt-remote-control and Gradle daemon took is interesting because they'd essentially launch something that'll stay permanent in the background.
sbt server reboot intentionally for the sake of simplicity (fewer moving parts).

Basically sbt docs have always documented sbt shell usages, so just updating docs might not be enough. Maybe we can print out couple lines of log saying something like:

```
[info] executing batch mode. note compilation is much faster if you ran from sbt shell mode.
```

**Comment by adriaanm** [comment]
This issue cuts across our whole stack. I like the idea of warning when sbt is run in single-task mode from the command line. 

Could we make it more "interactive" with a message like "[info] running task in batch mode; for better performance, hit any key to remain in the shell". Then, if they hit a key, run the `shell` command, and explain that next time they can just run `sbt` and then launch their task?

**Comment by eed3si9n** [comment]
> Could we make it more "interactive" with a message like...

You mean if someone runs `sbt test` on Travis CI or Jenkins, there's a JLine thread monitoring if the user hits any key? I feel like that's a risky move.

**Comment by leifwickland** [comment]
I agree that watching for input to sbt in batch mode is risky and wouldn't recommend it.

**Comment by adriaanm** [comment]
If you could reliably detect whether you're in interactive mode, you could pass that as an argument to sbt from the runner script. See https://github.com/scala/scala/pull/5663 for how we propose to do this for the scala runner.

**Comment by dwijnand** [comment]
I think such a feature could be hugely benefitial to sbt's reputation, so we should defintely not shy away and give it a proper go. We should definitely test various scenarios to harden the implementation (there's a bunch of shenanigans in the history of sbt-extras working out its `-batch` feature). And then we'll include an opt-out sys prop, as a fallback (and for any paranoid users).

**Comment by dwijnand** [comment]
The runtime feature request is being tracked at https://github.com/sbt/sbt/issues/2987.

**Comment by bennettellis** [comment]
Stupid question maybe, but is this real time savings? What I mean is in an earlier comment someone claimed "there's 2x difference in compiling from batch mode vs compiling from sbt shell." 

I'm not seeing it.

So for clarification, are you really just saying that compiling from an already-opened shell saves the time to open the shell (startup Java etc). If so, maybe just say exactly that.

For our project, I started a shell and not counting the startup time, here's the output of running clean, then compile.
[success] Total time: 9 s, completed Aug 30, 2017 8:15:24 AM
[success] Total time: 270 s, completed Aug 30, 2017 8:19:57 AM  
add about 5 seconds to start up the shell (type `sbt` wait for prompt)
Total time probably close to 300 s (5 minutes)

Then, for that same project, no changes, I ran `sbt clean compile`. Here's the output from running that:
[success] Total time: 253 s, completed Aug 30, 2017 8:25:54 AM  (4.2 minutes)

I'm not seeing the 2x difference here. In fact, I'm seeing slower compilation from an existing shell than running in batch mode. Maybe I'm misunderstanding or maybe I'm doing something different. Or maybe this is old and batch is more optimal these days, but it's annoying that I keep getting the message when I compile, but I don't get the real benefit claimed. Maybe the 2x is only realized if not clean?

Would love to be wrong here and find a faster compile time. Thanks!

**Comment by dwijnand** [comment]
The big difference in speed is running `compile` multiple times in an already-started and _warm_ sbt shell/JVM, compared to running `sbt compile` from the terminal all day. In the latter case you pay the startup and warmup cost every time.

**Comment by adriaanm** [comment]
In addition to the significant warmup cost, you're also running much slower code (interpreted or C1 compiled) until the JIT has done its thing and gotten to the C2 compiler. Other build tools (like gradle) hide the fact that you should keep these classes loaded by starting a compilation daemon in the background. (We should/will do this too, and then this message will become redundant.)
